pkg/providers/amifamily/ami_test.go

```diff
@@ -12,178 +12,346 @@ See the License for the specific language governing permissions and
+					{Key: aws.String("Name"), Value: aws.String(amd64NvidiaAMIName)},
+					{Key: aws.String("foo"), Value: aws.String("bar")},
+				},
 			},
-		}))
-	})
-	It("should not set owners when prefixed ids are passed", func() {
-		amiSelector := map[string]string{
-			"aws::ids": "ami-abcd1234,ami-cafeaced",
-		}
-		filters, owners := getFiltersAndOwners(amiSelector)
-		Expect(owners).Should(BeNil())
-		Expect(filters).Should(ConsistOf([]*ec2.Filter{
 			{
-				Name: aws.String("image-id"),
-				Values: aws.StringSlice([]string{
-					"ami-abcd1234",
-					"ami-cafeaced",
-				}),
+				Name:         aws.String(arm64NvidiaAMIName),
+				ImageId:      aws.String("arm64-nvidia-ami-id"),
+				CreationDate: aws.String(time.Now().Add(2 * time.Minute).Format(time.RFC3339)),
+				Architecture: aws.String("arm64"),
+				Tags: []*ec2.Tag{
+					{Key: aws.String("Name"), Value: aws.String(arm64NvidiaAMIName)},
+					{Key: aws.String("foo"), Value: aws.String("bar")},
+				},
 			},
-		}))
+		},
 	})
-	It("should allow only specifying owners", func() {
-		amiSelector := map[string]string{
-			"aws::owners": "abcdef,123456789012",
-		}
-		_, owners := getFiltersAndOwners(amiSelector)
-		Expect(owners).Should(ConsistOf(
-			[]*string{aws.String("abcdef"), aws.String("123456789012")},
-		))
+})
+
+var _ = AfterEach(func() {
+	awsEnv.Reset()
+})
+
+var _ = AfterSuite(func() {
+	Expect(env.Stop()).To(Succeed(), "Failed to stop environment")
+})
+
+var _ = Describe("AMI Provider", func() {
+	var version string
+	BeforeEach(func() {
+		version = lo.Must(awsEnv.AMIProvider.KubeServerVersion(ctx))
 	})
-	It("should allow prefixed id, prefixed name, and prefixed owners", func() {
-		amiSelector := map[string]string{
-			"aws::name":   "my-ami",
-			"aws::ids":    "ami-abcd1234,ami-cafeaced",
-			"aws::owners": "self,amazon",
-		}
-		filters, owners := getFiltersAndOwners(amiSelector)
-		Expect(owners).Should(ConsistOf(defaultOwners))
-		Expect(filters).Should(ConsistOf([]*ec2.Filter{
-			{
-				Name:   aws.String("name"),
-				Values: aws.StringSlice([]string{"my-ami"}),
-			},
-			{
-				Name: aws.String("image-id"),
-				Values: aws.StringSlice([]string{
-					"ami-abcd1234",
-					"ami-cafeaced",
-				}),
+	It("should succeed to resolve AMIs (AL2)", func() {
+		nodeTemplate = test.AWSNodeTemplate(v1alpha1.AWSNodeTemplateSpec{
+			AWS: v1alpha1.AWS{
+				AMIFamily: &v1alpha1.AMIFamilyAL2,
 			},
-		}))
+		})
+		awsEnv.SSMAPI.Parameters = map[string]string{
+			fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2/recommended/image_id", version):       amd64AMIName,
+			fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2-gpu/recommended/image_id", version):   amd64NvidiaAMIName,
+			fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2-arm64/recommended/image_id", version): arm64AMIName,
+		}
+		amis, err := awsEnv.AMIProvider.Get(ctx, nodeTemplate, &amifamily.Options{})
+		Expect(err).ToNot(HaveOccurred())
+		Expect(amis).To(HaveLen(4))
```

Comment:
Because there are 4 separate requirements sets that can't be combined. We can't have a single requirements set that says you can have "this or this" so they have to be separate requirements sets to fit the "OR" semantic

Reply:
See: https://github.com/aws/karpenter/blob/main/pkg/providers/amifamily/al2.go#L37